File: lab3/lab3-2.py
class Solution:
    def longestNiceSubstring(self, s):
        def is_nice(sub):
            chars = set(sub)
            for char in chars:
                if char.islower() and char.upper() not in chars:
                    return False
                if char.isupper() and char.lower() not in chars:
                    return False
            return True
        def longest_nice_substring(left, right):
            if left >= right:
                return ""

            chars = set(s[left:right + 1])
            for i in range(left, right + 1):
                if not (s[i].lower() in chars and s[i].upper() in chars):
                    left_part = longest_nice_substring(left, i - 1)
                    right_part = longest_nice_substring(i + 1, right)
                    return left_part if len(left_part) >= len(right_part) else right_part

            return s[left:right + 1]

        return longest_nice_substring(0, len(s) - 1)

if __name__ == "__main__":
    sol = Solution()
    s1 = "YazaAay"
    s2 = "Bb"
    s3 = "c"
    
    print(sol.longestNiceSubstring(s1))  
    print(sol.longestNiceSubstring(s2))  
    print(sol.longestNiceSubstring(s3))  

File: lab3/lab3-2.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <algorithm>
using namespace std;

class Solution {
public:
    string longestNiceSubstring(string s) {
        return longestNiceSubstringHelper(s, 0, s.size() - 1);
    }

private:
    bool isNice(const string& sub) {
        unordered_set<char> chars(sub.begin(), sub.end());
        for (char c : chars) {
            if (islower(c) && chars.find(toupper(c)) == chars.end()) {
                return false;
            }
            if (isupper(c) && chars.find(tolower(c)) == chars.end()) {
                return false;
            }
        }
        return true;
    }

    string longestNiceSubstringHelper(const string& s, int left, int right) {
        if (left >= right) {
            return "";
        }

        unordered_set<char> chars(s.begin() + left, s.begin() + right + 1);
        for (int i = left; i <= right; ++i) {
            if (!(chars.count(tolower(s[i])) && chars.count(toupper(s[i])))) {
                string leftPart = longestNiceSubstringHelper(s, left, i - 1);
                string rightPart = longestNiceSubstringHelper(s, i + 1, right);
                return leftPart.size() >= rightPart.size() ? leftPart : rightPart;
            }
        }

        return s.substr(left, right - left + 1);
    }
};

int main() {
    Solution sol;
    string s1 = "YazaAay";
    string s2 = "Bb";
    string s3 = "c";

    cout << sol.longestNiceSubstring(s1) << endl;  
    cout << sol.longestNiceSubstring(s2) << endl;  
    cout << sol.longestNiceSubstring(s3) << endl;  

    return 0;
}